scripts/run_portserver_test.py

```diff
@@ -0,0 +1,502 @@
+            result = run_portserver.should_allocate_port(pid)
+
+        self.assertFalse(result)
+        self.assertIn(
+            'Not allocating a port to invalid pid', self.terminal_logs)
+
+    def test_should_allocate_port_handles_init_pid(self) -> None:
+        pid = 1
+        with self.swap_log:
+            result = run_portserver.should_allocate_port(pid)
+
+        self.assertFalse(result)
+        self.assertIn(
+            'Not allocating a port to init.', self.terminal_logs)
+
+    def test_should_allocate_port_handles_oserror(self) -> None:
+        pid = 12345
+        def mock_kill(*unused_args: Any) -> None:
+            raise OSError('Some XYZ error occurred.')
+        swap_os_kill = self.swap_with_checks(
+            os, 'kill', mock_kill, expected_args=((pid, 0),))
+        with swap_os_kill, self.swap_log:
+            result = run_portserver.should_allocate_port(pid)
+
+        self.assertFalse(result)
+        self.assertIn(
+            'Not allocating a port to a non-existent process',
+            self.terminal_logs)
+
+    def test_port_pool_handles_invalid_port_request(self) -> None:
+        port = -1
+        port_pool = run_portserver.PortPool() # type: ignore[no-untyped-call]
+        error_msg = r'Port must be in the \[1, 65535\] range, not -1.'
+        with self.assertRaisesRegex(ValueError, error_msg):
+            port_pool.add_port_to_free_pool(port)
+
+    def test_port_pool_handles_empty_port_queue(self) -> None:
+        port_pool = run_portserver.PortPool() # type: ignore[no-untyped-call]
+        error_msg = 'No ports being managed.'
+        with self.assertRaisesRegex(RuntimeError, error_msg):
+            port_pool.get_port_for_process(12345)
+
+    def test_get_port_for_process_successfully(self) -> None:
+        port = 8181
+        swap_get_process_start_time = self.swap(
+            run_portserver, 'get_process_start_time', lambda _: 0)
+        swap_is_port_free = self.swap(
+            run_portserver, 'is_port_free', lambda _: True)
+
+        port_pool = run_portserver.PortPool() # type: ignore[no-untyped-call]
+        port_pool.add_port_to_free_pool(port)
+        self.assertEqual(port_pool.num_ports(), 1)
+        with swap_get_process_start_time, swap_is_port_free:
+            returned_port = port_pool.get_port_for_process(12345)
+
+        self.assertEqual(returned_port, port)
+
+    def test_get_port_for_process_looks_for_free_port(self) -> None:
+        port1 = 8181
+        port2 = 8182
+        swap_get_process_start_time = self.swap(
+            run_portserver, 'get_process_start_time', lambda _: 1)
+        swap_is_port_free = self.swap(
+            run_portserver, 'is_port_free', lambda _: True)
+
+        port_pool = run_portserver.PortPool() # type: ignore[no-untyped-call]
+        port_pool.add_port_to_free_pool(port1)
+        port_pool.add_port_to_free_pool(port2)
+        port = port_pool._port_queue.pop() # pylint: disable=protected-access
```

Comment:
ditto: no accessing private attributes

Reply:
Added a comment above in the test to describe why we are accessing private attributes here.